napari/layers/shapes/_shape_list.py

```diff
@@ -572,91 +545,373 @@ def _make_index(length, shape_index, cval=0):
+            Each Shape must be a subclass of Shape
 
-            triangles = shape._face_triangles + m_tmp
-            all_mesh_triangles.append(triangles)
-            all_mesh_triangles_index.append(
-                _make_index(len(triangles), shape_index, cval=0)
+        Returns
+        -------
+        n_shapes : int
+            Number of shapes
+        total_vertices_data : int
+            Total number of vertices
+        total_index_data : int
+            Total number of indices
+        total_mesh_vertices : int
+            Total number of mesh vertices
+        total_face_triangles : int
+            Total number of face triangles
+        total_edge_triangles : int
+            Total number of edge triangles
+        """
+        n_shapes = len(shapes)
+        total_vertices_data = 0
+        total_index_data = 0
+        total_mesh_vertices = 0
+        total_face_triangles = 0
+        total_edge_triangles = 0
+
+        for shape in shapes:
+            total_vertices_data += len(shape.data_displayed)
+            total_index_data += len(shape.data)
+            total_mesh_vertices += len(shape._face_vertices) + len(
+                shape._edge_vertices
             )
+            total_face_triangles += len(shape._face_triangles)
+            total_edge_triangles += len(shape._edge_triangles)
+
+        return (
+            n_shapes,
+            total_vertices_data,
+            total_index_data,
+            total_mesh_vertices,
+            total_face_triangles,
+            total_edge_triangles,
+        )
 
-            color_array = np.repeat([face_color], len(triangles), axis=0)
-            all_mesh_triangles_colors.append(color_array)
+    def _preallocate_arrays(self, shapes, sizes):
+        """Preallocate arrays for storing shape data.
 
-            # Add edges to mesh
-            m_tmp = m_mesh_vertices_count
+        Parameters
+        ----------
+        shapes : iterable of Shape
+            Each Shape must be a subclass of Shape
+        sizes : tuple
+            Tuple containing sizes for preallocation:
+            (n_shapes, total_vertices_data, total_index_data, total_mesh_vertices,
+            total_face_triangles, total_edge_triangles)
 
-            vertices = (
-                shape._edge_vertices + shape.edge_width * shape._edge_offsets
-            )
-            all_mesh_vertices.append(vertices)
-            m_mesh_vertices_count += len(vertices)
+        Returns
+        -------
+        arrays : dict
+            Dictionary containing preallocated arrays
+        """
+        (
+            n_shapes,
+            total_vertices_data,
+            total_index_data,
+            total_mesh_vertices,
+            total_face_triangles,
+            total_edge_triangles,
+        ) = sizes
+
+        # Determine the dimensionality from the first shape
+        # All shapes in a batch will have the same dimensionality
+        dim = shapes[0]._face_vertices.shape[1] if shapes else 3
```

Comment:
Why is the default dimensionality 3?

Reply:
And is this the same as self.ndisplay?

Reply:
I'm working on minor refactor currently so I can update, just wanted to understand

Reply:
still the empty arrays should be created, no?

Reply:
... maybe not

Reply:
No, we can return before we even do anything